src/promptflow-devkit/promptflow/batch/_batch_engine.py

```diff
@@ -545,39 +545,64 @@ async def _exec_line_under_semaphore(
         async with semaphore:
             return await self._executor_proxy.exec_line_async(inputs, index, run_id)
 
-    async def _exec_aggregation(
-        self,
-        batch_inputs: List[dict],
-        line_results: List[LineResult],
-        run_id: Optional[str] = None,
-    ) -> AggregationResult:
+    def _should_exec_aggregation(self) -> bool:
         if self._is_eager_flow:
-            return AggregationResult({}, {}, {})
+            return self._executor_proxy.has_aggregation
         aggregation_nodes = {node.name for node in self._flow.nodes if node.aggregation}
-        if not aggregation_nodes:
-            return AggregationResult({}, {}, {})
-
-        bulk_logger.info("Executing aggregation nodes...")
+        return bool(aggregation_nodes)
 
+    def _get_aggregation_inputs(self, batch_inputs, line_results: List[LineResult]):
         run_infos = [r.run_info for r in line_results]
         succeeded = [i for i, r in enumerate(run_infos) if r.status == Status.Completed]
 
+        if self._is_eager_flow:
+            return None, [line_results[i].output for i in succeeded]
+
         succeeded_batch_inputs = [batch_inputs[i] for i in succeeded]
         resolved_succeeded_batch_inputs = [
             FlowValidator.ensure_flow_inputs_type(flow=self._flow, inputs=input) for input in succeeded_batch_inputs
         ]
-
         succeeded_inputs = transpose(resolved_succeeded_batch_inputs, keys=list(self._flow.inputs.keys()))
-
         aggregation_inputs = transpose(
             [result.aggregation_inputs for result in line_results],
             keys=get_aggregation_inputs_properties(self._flow),
         )
         succeeded_aggregation_inputs = collect_lines(succeeded, aggregation_inputs)
+        return succeeded_inputs, succeeded_aggregation_inputs
+
+    async def _exec_aggregation(
+        self,
+        batch_inputs: List[dict],
+        line_results: List[LineResult],
+        run_id: Optional[str] = None,
+    ) -> AggregationResult:
+        if not self._should_exec_aggregation():
+            return AggregationResult({}, {}, {})
+
+        bulk_logger.info("Executing aggregation nodes...")
+
+        inputs, aggregation_inputs = self._get_aggregation_inputs(batch_inputs, line_results)
+
+        if self._is_eager_flow:
```

Comment:
We should not have such if/else logic here, those error handling logic should be put in executor.

Reply:
Refined.